esphome/components/emmeti/emmeti.cpp

```diff
@@ -0,0 +1,316 @@
+        break;
+      case EMMETI_BLADES_4:
+      case EMMETI_BLADES_5:
+      case EMMETI_BLADES_LOW:
+        this->blades_ = EMMETI_BLADES_LOW;
+        break;
+      default:
+        this->blades_ = EMMETI_BLADES_FULL;
+        break;
+    }
+  } else {
+    switch (this->blades_) {
+      case EMMETI_BLADES_1:
+      case EMMETI_BLADES_2:
+      case EMMETI_BLADES_HIGH:
+        this->blades_ = EMMETI_BLADES_1;
+        break;
+      case EMMETI_BLADES_3:
+      case EMMETI_BLADES_MID:
+        this->blades_ = EMMETI_BLADES_3;
+        break;
+      case EMMETI_BLADES_4:
+      case EMMETI_BLADES_5:
+      case EMMETI_BLADES_LOW:
+        this->blades_ = EMMETI_BLADES_5;
+        break;
+      default:
+        this->blades_ = EMMETI_BLADES_STOP;
+        break;
+    }
+  }
+  return this->blades_;
+}
+
+uint8_t EmmetiClimate::gen_checksum_() { return (this->set_temp_() + this->set_mode_() + 2) % 16; }
+
+// getters
+float EmmetiClimate::get_temp_(uint8_t temp) { return (float) (temp + EMMETI_TEMP_MIN); }
+
+climate::ClimateMode EmmetiClimate::get_mode_(uint8_t mode) {
+  switch (mode) {
+    case EMMETI_MODE_COOL:
+      return climate::CLIMATE_MODE_COOL;
+    case EMMETI_MODE_DRY:
+      return climate::CLIMATE_MODE_DRY;
+    case EMMETI_MODE_HEAT:
+      return climate::CLIMATE_MODE_HEAT;
+    case EMMETI_MODE_HEAT_COOL:
+      return climate::CLIMATE_MODE_HEAT_COOL;
+    case EMMETI_MODE_FAN:
+      return climate::CLIMATE_MODE_FAN_ONLY;
+    default:
+      return climate::CLIMATE_MODE_HEAT_COOL;
+  }
+}
+
+climate::ClimateFanMode EmmetiClimate::get_fan_speed_(uint8_t fan_speed) {
+  switch (fan_speed) {
+    case EMMETI_FAN_1:
+      return climate::CLIMATE_FAN_LOW;
+    case EMMETI_FAN_2:
+      return climate::CLIMATE_FAN_MEDIUM;
+    case EMMETI_FAN_3:
+      return climate::CLIMATE_FAN_HIGH;
+    case EMMETI_FAN_AUTO:
+    default:
+      return climate::CLIMATE_FAN_AUTO;
+  }
+}
+
+climate::ClimateSwingMode EmmetiClimate::get_swing_(uint8_t bitmap) {
+  return (bitmap >> 1) & 0x01 ? climate::CLIMATE_SWING_VERTICAL : climate::CLIMATE_SWING_OFF;
+}
+
+template<typename T> T EmmetiClimate::reverse_(T val, size_t len) {
+  T result = 0;
+  for (size_t i = 0; i < len; i++) {
+    result |= ((val & 1 << i) != 0) << len - 1 - i;
+  }
+  return result;
+}
+
+template<typename T> void EmmetiClimate::add_(T val, size_t len, esphome::remote_base::RemoteTransmitData *data) {
+  for (size_t i = len; i > 0; i--) {
+    data->mark(EMMETI_BIT_MARK);
+    data->space((val & (1 << (i - 1))) ? EMMETI_ONE_SPACE : EMMETI_ZERO_SPACE);
+  }
+}
+
+template<typename T> void EmmetiClimate::add_(T val, esphome::remote_base::RemoteTransmitData *data) {
+  data->mark(EMMETI_BIT_MARK);
+  data->space((val & 1) ? EMMETI_ONE_SPACE : EMMETI_ZERO_SPACE);
+}
+
+template<typename T>
+void EmmetiClimate::reverse_add_(T val, size_t len, esphome::remote_base::RemoteTransmitData *data) {
+  add_(reverse_(val, len), len, data);
```

Comment:
Please ensure all class members and member functions are prefixed with `this->` 
```suggestion
  this->add_(reverse_(val, len), len, data);
```